packages/access-api/src/service/provider-add.js

```diff
@@ -0,0 +1,58 @@
+import * as Ucanto from '@ucanto/interface'
+import * as Server from '@ucanto/server'
+import { Provider } from '@web3-storage/capabilities'
+import * as validator from '@ucanto/validator'
+
+/**
+ * @typedef {import('@web3-storage/capabilities/types').ProviderAdd} ProviderAdd
+ * @typedef {import('@web3-storage/capabilities/types').ProviderAddSuccess} ProviderAddSuccess
+ * @typedef {import('@web3-storage/capabilities/types').ProviderAddFailure} ProviderAddFailure
+ */
+
+/**
+ * @callback ProviderAddHandler
+ * @param {Ucanto.Invocation<import('@web3-storage/capabilities/types').ProviderAdd>} invocation
+ * @returns {Promise<Ucanto.Result<ProviderAddSuccess, ProviderAddFailure>>}
+ */
+
+/**
+ * @param {object} options
+ * @param {import('../types/provisions').StorageProvisions} options.storageProvisions
+ * @returns {ProviderAddHandler}
+ */
+export function createProviderAddHandler(options) {
```

Comment:
Nit: This obviously follows the style of code that is in the repo already, but I do find amount of indirection adds so much unnecessary complexity e.g. I would much rather have something like:

```ts
/**
 * @param {object} input
 * @param {Ucanto.Invocation<import('@web3-storage/capabilities/types').ProviderAdd>} input.invocation
 * @param {{ storageProvisions: import('../types/provisions').StorageProvisions }} input.context
 */
export const add ({ invocation, context: { storageProvisions } }) => {
   // ...
}
```

And in the service defs something like

```js
Server.provide(Provider.add, ({ invocation }) => add({ invocation, context })
```

Than all this layers and closures all over.

Also for what it's worth I'm going to add `context` to the ucanto provider handles to remove the need for all those closures.

Reply:
this is a good idea and I'll try it out across all the relevant things I wrote in that style to use the less-closures style you suggest. https://github.com/web3-storage/w3protocol/issues/481